Add support for HTML bold/italic/font size and font family to labelling

nyalldawson opened the pull request:
This PR adds support for additional HTML formatting tags within labeling (and everywhere text renderer is used). Specifically:

- font family
- font size
- bold
- italic

Sponsored by the Swiss QGIS user group



Fixes https://github.com/qgis/QGIS/issues/40532
Fixes https://github.com/qgis/QGIS/issues/32098
Fixes https://github.com/qgis/QGIS/issues/18727
Fixes https://github.com/qgis/QGIS/issues/31501

nyalldawson committed:
Add font size, family, italic and bold to QgsTextCharacterFormat

nyalldawson committed:
Fix incorrect argument passed to function

(should have no impact on existing releases, as the text fragements
previously did not support properties which altered font metrics)

nyalldawson committed:
Minor cleanup

nyalldawson committed:
Correctly account for scaling factor and render context scale
when creating fonts for document fragments

nyalldawson committed:
Modify text renderer to account for rendering text with HTML
content which varies in font metrics

nyalldawson committed:
Correctly account for text sizes in curved labels

nyalldawson committed:
Fix qt 6 build

nirvn commented:
@nyalldawson , did some testing, it's really nice, good job.

I did run into a crash (having to do with curved labels I think):
```
#6  0x000055555556567b in myMessageOutput(QtMsgType, QMessageLogContext const&, QString const&)
    (type=QtFatalMsg, msg=...) at /home/webmaster/dev/cpp/QGIS/src/app/main.cpp:455
#7  0x00007ffff3a2e530 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#8  0x00007ffff3a2e649 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ffff39fab8a in QMessageLogger::fatal(char const*, ...) const ()
--Type <RET> for more, q to quit, c to continue without paging--
   .5
#10 0x00007ffff39f9ff0 in qt_assert_x(char const*, char const*, char const*, int) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff142a73d in QList<double>::last() const (this=0x7fffcbffb378) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qlist.h:365
#12 0x00007ffff142a134 in QList<double>::constLast() const (this=0x7fffcbffb378) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qlist.h:366
#13 0x00007ffff1429a41 in QgsTextDocumentMetrics::calculateMetrics(QgsTextDocument const&, QgsTextFormat const&, QgsRenderContext const&, double)
    (document=..., format=..., context=..., scaleFactor=1) at /home/webmaster/dev/cpp/QGIS/src/core/textrenderer/qgstextdocumentmetrics.cpp:153
#14 0x00007ffff1e0150b in QgsPalLayerSettings::calculateLabelSize(QFontMetricsF const*, QString const&, double&, double&, QgsFeature const*, QgsRenderContext*, double*, double*, QgsTextDocument*, QgsTextDocumentMetrics*)
    
     (this=0x55556062e520, fm=0x7fffc07238c0, text=..., labelX=@0x7fffcbffb4e8: 2.5599981057214506e-172, labelY=@0x7fffcbffb4f0: 9.8813129168249309e-324, f=0x7fffcbffbaa0, context=0x55556062b5d0, rotatedLabelX=0x7fffcbffb4f8, rotatedLabelY=0x7fffcbffb500, document=0x7fffcbffb508, documentMetrics=0x7fffcbffb890) at /home/webmaster/dev/cpp/QGIS/src/core/labeling/qgspallabeling.cpp:1630
#15 0x00007ffff1e03f30 in QgsPalLayerSettings::registerFeatureWithDetails(QgsFeature const&, QgsRenderContext&, QgsGeometry, QgsSymbol const*)
    (this=0x55556062e520, f=..., context=..., obstacleGeometry=..., symbol=0x55556062c2d0)
    at /home/webmaster/dev/cpp/QGIS/src/core/labeling/qgspallabeling.cpp:2025
#16 0x00007ffff1e2e9f4 in QgsVectorLayerLabelProvider::registerFeature(QgsFeature const&, QgsRenderContext&, QgsGeometry const&, QgsSymbol const*)
    (this=0x55556062e4c0, feature=..., context=..., obstacleGeometry=..., symbol=0x55556062c2d0)
    at /home/webmaster/dev/cpp/QGIS/src/core/labeling/qgsvectorlayerlabelprovider.cpp:199
#17 0x00007ffff1fe3db2 in QgsVectorLayerRenderer::drawRenderer(QgsFeatureRenderer*, QgsFeatureIterator&)
     (this=0x55556062bf10, renderer=0x55556062d7a0, fit=...) at /home/webmaster/dev/cpp/QGIS/src/core/vector/qgsvectorlayerrenderer.cpp:511
#18 0x00007ffff1fe3039 in QgsVectorLayerRenderer::renderInternal(QgsFeatureRenderer*) (this=0x55556062bf10, renderer=0x55556062d7a0)
    at /home/webmaster/dev/cpp/QGIS/src/core/vector/qgsvectorlayerrenderer.cpp:409
#19 0x00007ffff1fe1a6c in QgsVectorLayerRenderer::render() (this=0x55556062bf10)
    at /home/webmaster/dev/cpp/QGIS/src/core/vector/qgsvectorlayerrenderer.cpp:225
```

I haven't worked out a reduced test project to replicate crash, but on a large project, I can replicate crash as soon as curved labels are on the canvas.

nyalldawson committed:
Ensure curved label text of mixed heights is vertically baseline aligned

In future we probably want to support explicit control of vertical
alignment for text, but for now baseline alignment is the best
default option we can use.

nyalldawson committed:
Fix bold html formatting is not always respected

nyalldawson committed:
Correctly place text backgrounds when valign isn't top

nyalldawson committed:
Correctly render vertical oriented text using HTML formatting

nyalldawson committed:
Don't crash on empty label text

nyalldawson committed:
Add unit tests

nyalldawson committed:
Doxygen

nyalldawson committed:
Fix warning

nyalldawson committed:
Fix qt6 build

nirvn commented on src/core/labeling/qgstextlabelfeature.h:
```diff
@@ -121,11 +110,19 @@ class CORE_EXPORT QgsTextLabelFeature : public QgsLabelFeature
+    QgsTextDocumentMetrics documentMetrics() const;
+
+    /**
+     * Sets the \a document and document \a metrics for the label.
```
Should we \note the metrics parameter appeared in QGIS 3.28? No strong feelings, you can ignore.

nirvn commented on src/core/textrenderer/qgstextcharacterformat.h:
```diff
@@ -162,19 +203,16 @@ class CORE_EXPORT QgsTextCharacterFormat
      * at small font sizes.
      */
-    void updateFontForFormat( QFont &font, double scaleFactor = 1.0 ) const;
+    void updateFontForFormat( QFont &font, const QgsRenderContext &context, double scaleFactor = 1.0 ) const;
```
Should we document the \a context parameter here?

nirvn commented on src/core/textrenderer/qgstextdocumentmetrics.cpp:
```diff
@@ -0,0 +1,274 @@
+  // for absolute line heights
+  const double lineHeightPainterUnits = context.convertToPainterUnits( format.lineHeight(), format.lineHeightUnit() );
+
+  // TODO orientation handling
```
Is it still a TODO?

nirvn commented on src/core/textrenderer/qgstextdocumentmetrics.h:
```diff
@@ -0,0 +1,120 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSTEXTDOCUMENTMETRICS_H
```
No #pragma once ? :wink:

nirvn commented on src/core/textrenderer/qgstextmetrics.h:
```diff
@@ -63,6 +71,26 @@ class CORE_EXPORT QgsPrecalculatedTextMetrics
+    /**
+     * Returns the character height of the character at the specified position (actually font metrics height, not individual character height).
+     */
+    double characterHeight( int position ) const { return mCharacterHeights[position]; }
```
\since 3.28? (for this and the rest of them)

nirvn reviewed (commented):
The code change is, well, dense! :) I couldn't find anything wrong but it's not the easiest part of the src tree to navigate. I however did a fair amount of testing locally, and couldn't spot any further regressions. Good job. Just a few Qs.

nyalldawson committed:
Address review
